File: core/src/main/java/vn/loitp/views/textview/countdown/lib/LCountDownView.java
package vn.loitp.views.textview.countdown.lib;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.daimajia.androidanimations.library.Techniques;

import loitp.core.R;
import vn.loitp.core.utilities.LAnimationUtil;
import vn.loitp.core.utilities.LLog;

/**
 * Created by www.muathu@gmail.com on 5/13/2017.
 */

public class LCountDownView extends RelativeLayout {
    private final String TAG = getClass().getSimpleName();
    private TextView tvCountDown;

    public LCountDownView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public LCountDownView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_l_count_down, this);
        this.tvCountDown = (TextView) findViewById(R.id.tv_count_down);
    }

    public void start(int number) {
        this.number = number;
        doPerSec();
    }

    public void setShowOrHide(boolean isShow) {
        if (tvCountDown != null) {
            tvCountDown.setVisibility(isShow ? VISIBLE : INVISIBLE);
        }
    }

    private int number;

    public interface Callback {
        public void onTick();

        public void onEnd();
    }

    private Callback callback;

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    private void doPerSec() {
        if (tvCountDown != null) {
            tvCountDown.setText(number + "");
            LAnimationUtil.playDuration(tvCountDown, Techniques.FlipInX, 1000, new LAnimationUtil.Callback() {
                @Override
                public void onCancel() {
                    //do nothing
                }

                @Override
                public void onEnd() {
                    number--;
                    //LLog.d(TAG, "number " + number);
                    if (number == 0) {
                        tvCountDown.setText("GO");
                        LAnimationUtil.play(tvCountDown, Techniques.Flash, new LAnimationUtil.Callback() {
                            @Override
                            public void onCancel() {
                                //do nothing
                            }

                            @Override
                            public void onEnd() {
                                tvCountDown.setVisibility(GONE);
                                //tvCountDown = null;
                                LLog.d(TAG, "number == 0 -> STOP");
                                if (callback != null) {
                                    callback.onEnd();
                                }
                            }

                            @Override
                            public void onRepeat() {
                                //do nothing
                            }

                            @Override
                            public void onStart() {
                                //do nothing
                            }
                        });
                        return;
                    }
                    doPerSec();
                }

                @Override
                public void onRepeat() {
                    //do nothing
                }

                @Override
                public void onStart() {
                    //do nothing
                    if (callback != null) {
                        callback.onTick();
                    }
                }
            });
        }
    }
}

File: core/src/main/java/com/loitpcore/views/textView/countDown/LCountDownView.kt
package com.loitpcore.views.textView.countDown

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import com.daimajia.androidanimations.library.Techniques
import com.loitpcore.R
import com.loitp.core.utilities.LAnimationUtil
import kotlinx.android.synthetic.main.view_l_count_down.view.*

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class LCountDownView : RelativeLayout {

    interface Callback {
        fun onTick()
        fun onEnd()
    }

    private var number = 0
    private var callback: Callback? = null

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init()
    }

    private fun init() {
        inflate(context, R.layout.view_l_count_down, this)
    }

    fun start(number: Int) {
        this.number = number
        doPerSec()
    }

    fun setShowOrHide(isShow: Boolean) {
        tvCountDown.visibility = if (isShow) {
            VISIBLE
        } else {
            INVISIBLE
        }
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    @SuppressLint("SetTextI18n")
    private fun doPerSec() {
        tvCountDown.text = number.toString()
        LAnimationUtil.play(
            view = tvCountDown,
            techniques = Techniques.FlipInX,
            duration = 1000,
            onEnd = {
                number--
                if (number <= 0) {
                    tvCountDown.text = "GO"
                    LAnimationUtil.play(
                        view = tvCountDown,
                        techniques = Techniques.Flash,
                        onEnd = {
                            tvCountDown?.visibility = GONE
                            callback?.onEnd()
                        }
                    )
                } else {
                    doPerSec()
                }
            }
        )
    }
}
